internal/endpoint/pop3/pop3.go

```diff
@@ -0,0 +1,466 @@
+// Update() is called. Be aware that after Dele() is called, functions like List() etc.
+// should ignore all these messages even if Update() hasn't been called yet
+func (endp *Endpoint) Dele(user pop3backend.User, msgId int) error {
+	mailbox, err := endp.getMailbox(user)
+	if err != nil {
+		return err
+	}
+
+	seqset := imap.SeqSet{}
+	seqset.AddNum(uint32(msgId))
+	return mailbox.UpdateMessagesFlags(true, &seqset, imap.SetFlags, false, []string{imap.DeletedFlag})
+}
+
+// Undelete all messages marked as deleted in single connection
+func (endp *Endpoint) Rset(user pop3backend.User) error {
+	return fmt.Errorf("pop3: unimplemented")
+}
+
+// List of unique IDs of all message, similar to List(), but instead of size there
+// is a unique ID which persists the same across all connections. Uid (unique id) is
+// used to allow client to be able to keep messages on the server.
+func (endp *Endpoint) Uidl(user pop3backend.User) (uids []string, err error) {
+	mailbox, err := endp.getMailbox(user)
+	if err != nil {
+		return nil, err
+	}
+
+	c := make(chan *imap.Message)
+	err = mailbox.ListMessages(false, nil, nil, c)
+	if err != nil && err != mess.ErrNoMessages {
+		return nil, err
+	}
+
+	items := make([]string, 0)
+	for msg := range(c) {
+		items = append(items, strconv.FormatUint(uint64(msg.Uid), 10))
+	}
+
+	return items, nil
+}
+
+// Similar to ListMessage, but returns unique ID by message ID instead of size.
+func (endp *Endpoint) UidlMessage(user pop3backend.User, msgId int) (exists bool, uid string, err error) {
+	mailbox, err := endp.getMailbox(user)
+	if err != nil {
+		return false, "", err
+	}
+
+	seqset := imap.SeqSet{}
+	seqset.AddNum(uint32(msgId))
+	c := make(chan *imap.Message)
+	err = mailbox.ListMessages(true, &seqset, nil, c)
+	if err != nil && err != mess.ErrNoMessages {
+		return false, "", err
+	}
+
+	msg, ok := <- c
+	if !ok {
+		return false, "", nil
+	}
+
+	return true, strconv.FormatUint(uint64(msg.Uid), 10), nil
+}
+
+// Write all changes to persistent storage, i.e. delete all messages marked as deleted.
+func (endp *Endpoint) Update(user pop3backend.User) error {
+	mailbox, err := endp.getMailbox(user)
+	if err != nil {
+		return err
+	}
+
+	return mailbox.Expunge()
+}
+
+// If the POP3 server issues a positive response, then the
+// response given is multi-line.  After the initial +OK, the
+// POP3 server sends the headers of the message, the blank
+// line separating the headers from the body, and then the
+// number of lines of the indicated message's body, being
+// careful to byte-stuff the termination character (as with
+// all multi-line responses).
+// Note that if the number of lines requested by the POP3
+// client is greater than than the number of lines in the
+// body, then the POP3 server sends the entire message.
+func (endp *Endpoint) Top(user pop3backend.User, msgId int, n int) (lines []string, err error) {
+	return nil, fmt.Errorf("pop3: unimplemented")
```

Comment:
Is there code I can reference for formatting the message headers?

Reply:
Hmm, that one is tricky. I guess, you would need to consume Literals returned by ListMessages and format them yourself. It is okay to partially consume returned value.